&nbsp;
<div style="text-align: center;">

## <span style="color: black;">Website Restoparts - restoration parts</span>

&nbsp;
<center><img src="restoparts_logo.png"/></center>


&nbsp;

### <span style="color: black;">Data Cleaning</span>

#### <span style="color: #003E99;">804 classic car body parts</span>

## Index

- [<font color='black'> 1. Packages installation</font>](#1)



- [<font color='black'> 2. Data importation</font>](#2)



- [<font color='black'> 3. Data exploration</font>](#3)
    - [<font color='black'> 3.1. Missing values and duplicated data</font>](#3.1)
    - [<font color='black'> 3.2. Most frequent itens</font>](#3.2)
    - [<font color='black'> 3.3. Name feature </font>](#3.3)
    - [<font color='black'> 3.4. Product type feature </font>](#3.4)
    - [<font color='black'> 3.5. Filter feature </font>](#3.5)
    - [<font color='black'> 3.6. Composition </font>](#3.6)
    - [<font color='black'> 3.7. Finish </font>](#3.7)
    - [<font color='black'> 3.8. Dimensions </font>](#3.8)
    - [<font color='black'> 3.9. Models </font>](#3.9)
    - [<font color='black'> 3.10. Price </font>](#3.10) 
 
&nbsp;

- [<font color='black'> 4. Data cleaning</font>](#4)
    - [<font color='black'> 4.1. Standardization of formats</font>](#4.1)
    - [<font color='black'> 4.2. Noise removal</font>](#4.2)
    - [<font color='black'> 4.3. Data type conversion</font>](#4.3)
      
&nbsp;

- [<font color='black'> 5. Data transformation</font>](#5)
        


- [<font color='black'> 6. Data integration</font>](#6)
    - [<font color='black'> 6.1. Models table</font>](#6.1)
    - [<font color='black'> 6.2. Composition and finish tables</font>](#6.2)


&nbsp;

- [<font color='black'> 7. Data selection</font>](#7)



- [<font color='black'> 8. Data exportation</font>](#8)


      


## 1. Packages installation
<a id='1'>

In [1]:
#!pip3 install unicode

In [2]:
#dataframes
import pandas as pd
import numpy as np

# compativel com linguagem ASCII - unicode
from unidecode import unidecode 

#Expressões regulares
import re


## 2. Data importation
<a id='2'>

In [3]:
db = pd.read_csv('Bodyworkparts_rawdata.csv', sep=',')

## 3. Data exploration
<a id='3'>

In [4]:
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [5]:
db.head(2)

Name  \
0     DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)   
1  SCREEN, COWL VENT, 1968-72 A-BODY, 1969-72 GRAND PRIX, W/ AC   

                       Product_type               Composition   Finish  \
0  Hardware, Fasteners and Fittings             Stamped steel  Natural   
1         Cowl Induction Components  Injection-molded plastic    Black   

    Weight   Length   Width  Height  \
0  0.2 lbs   5.0 in  3.5 in  0.2 in   
1  0.3 lbs  14.8 in  4.8 in  1.5 in   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Models  \
0  ['1967-1976 Buick Riviera', '1967-1972 Buick Skylark', '1967-1977 Cadillac DeVille', '1967-1977 Cadillac Eldorado', '1967-1977 Cadillac Fleetwood', '1967-1981 Chevrolet Camaro', '1967-1977 Chevrolet Chevelle', '1967-1977 Chevrolet El Camino', '1964-1972 Chevrolet Impala', '1970-1977 Chevrolet Monte Carlo', '1962-1972 Chevrolet Nova', '1971-1972 GMC Sprint', '1967-1977 Oldsmobile Cutlass', '1967-1977 Pontiac Bonneville', '1967-1977 Pontiac Catalina', '1967-1981 Pontiac Firebird', '1967-1977 Pontiac Grand Prix', '1967-1973 Pontiac GTO', '1967-1973 Pontiac LeMans', '1967-1970 Pontiac Tempest']   
1                                                                                                                                                                                                                                                                                                                                      ['1968-1972 Buick Skylark', '1968-1972 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1968-1972 Oldsmobile Cutlass', '1969-1972 Pontiac Grand Prix', '1968-1972 Pontiac GTO', '1968-1972 Pontiac LeMans', '1968-1970 Pontiac Tempest']   

          Price  
0   MSRP: $6.99  
1  MSRP: $26.99

<span style="color:blue">

- the name is too expensive
- lbs should be removed from weight values and added to the weight column title such as 'weight (lbs)'
- in should be removed and added to Lenght, Width, and Height column titles such as 'Lenght (in)
- should Models have another organization? such as a relational table for each model? in phpmyadmin and consulted in MongoDB?
- price without MSRP: $
  

### 3.1. Missing values and duplicated data
<a id='3.1'>

In [6]:
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 804 entries, 0 to 803
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Name          804 non-null    object
 1   Product_type  767 non-null    object
 2   Composition   746 non-null    object
 3   Finish        665 non-null    object
 4   Weight        804 non-null    object
 5   Length        804 non-null    object
 6   Width         804 non-null    object
 7   Height        804 non-null    object
 8   Models        804 non-null    object
 9   Price         804 non-null    object
dtypes: object(10)
memory usage: 62.9+ KB


<span style="color:blue">

- Weight, Length, Width, Height should be **numeric variables**

    
| Variable     | NaN |
|:-----------------|-----------------------|
| Product_type      | 37                 |
| Composition      | 58                |
| Finish    | 139               |
|


<span style="color:blue">

- Only product type, composition, and finish features have missing values
- Finish has a lot of missing values, one eighth of the dataset

In [7]:
# Duplicated data

print(db.duplicated().sum())

0


<span style="color:blue">

- no duplicated data

In [8]:
#db = db.drop_duplicates

### 3.2. Most frequent itens
<a id='3.2'>

In [9]:
db.describe()

Name  \
count                                                         804   
unique                                                        804   
top     DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)   
freq                                                            1   

          Product_type Composition      Finish   Weight  Length   Width  \
count              767         746         665      804     804     804   
unique              80          40          35      196     247     178   
top     Quarter Panels       Steel  EDP coated  4.5 lbs  0.0 in  0.0 in   
freq               108         487         358       27      39      39   

        Height                            Models          Price  
count      804                               804            804  
unique     102                               222            291  
top     2.0 in  ['1970-1972 Chevrolet Chevelle']  MSRP: $119.99  
freq        50                                42             28

<span style="color:blue">

- most of the parts are quarter panels, one eight of the dataset, reaching 108 parts

- most of the parts are of steel composition, with 487 parts being made of steel
- most of the parts have a finishing with EDP coat, about 358 parts with this material

- the classic car with the most parts available on this website is the Chevrolet Chevelle of 1970-1972???

### 3.3. Name feature
<a id='3.3'>

In [10]:
unique_name = db['Name'].unique().tolist()
print(unique_name)

['DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)', 'SCREEN, COWL VENT, 1968-72 A-BODY, 1969-72 GRAND PRIX, W/ AC', 'SPRING, HOOD HINGE, 1964-67/1970-72 A-BODY', '1964-1977 HOOD ADJUSTER BOLT KIT, GM A-BODY, RIVIERA, 1959-1963 BONNEVILLE AND CATALINA', 'HOOD PIN SET, 1970-72 CHEVELLE/EL CAMINO, SS', 'SEAL, INNER FENDERWELL A-ARM, 1968-72 CHEVELLE/EL CAMINO/MONTE CARLO, STEEL', 'SCREEN, COWL VENT, 1968-72 W/O AC', 'FIREWALL WIRE ROUTING GUTTER', '1954-1977 GM 7/16" X 14 CAGE NUT', 'DRAIN PLUG, FLOOR & TRUNK PAN, 1961-67 GM, NOTCHED', 'FASTENER SET, FRONT END, 1968-72 CHEVELLE/EL CAMINO/MONTE CARLO', 'INSULATION PAD, FIREWALL, 1964-67 A-BODY, W/AC', 'INSULATION PAD, FIREWALL, 1968-72 A-BODY, NO AC', 'HOOD HINGE, 1970-72 CHEVELLE/EL CAMINO/MONTE CARLO, REPRO', 'BOLT, FENDER TO BODY, 1978-87 G-BODY', 'STOP CLIP, WINDSHIELD GLASS, 1968-72 A-BODY', 'HARDWARE, INNER FENDER, 1964-72 CHEV/EL CAM/MONTE/GTO/ LEM/TEMP', 'EMBLEM, HOOD, 1970-72 CHEVELLE/EL CAMINO, SS, COWL INDUCTION', 'HOO

### 3.4. Product type feature
<a id='3.4'>

In [11]:
unique_pt = db['Product_type'].unique().tolist()
print(unique_pt)

['Hardware, Fasteners and Fittings', 'Cowl Induction Components', 'Hood', 'Bolts & Studs', 'Hood Components', 'Gaskets and Sealing Systems', 'Hardware', 'Dash & Accessories', 'Fenders', 'Window Component', 'Exterior Emblems', 'Rear Body', nan, 'Hood Component', 'Tailgate and Midgate', 'Floor Component', 'Accessory Bolt Kits', 'Air Cleaners, Flame Arrestors & Scoops', 'Cowl Induction', 'Exterior Locks & Hardware', 'Engine Components', 'Engine Component', 'Tailgate Components', 'Fender Components', 'Cowl induction', 'Air & Fuel Delivery', 'Floor Pans', 'Body Component', 'Trunk Lid and Compartment', 'Trunk Panels', 'Cowl', 'Trunk Component', 'Roof Component', 'Chassis Components', 'Doors', 'Quarter Panels', 'Firewall/Cowl', 'Tailgate Component', 'Toe Board', 'Valance', 'Body & Trunk Braces', 'Quarter Panel', 'Bumper Component', 'Rear/Tail End', 'Bumper Panel', 'Rocker Panel', 'Sheet metal components', 'Radiators, Coolers and Related Components', 'Aero Ground Effects and Body Kits', 'Front

<span style="color:blue">

- choose wich one are near of the parts investigated from Raimundo company, relative to:

- tejadilho ------------------- **Roof, Roof Component**
- capo/bagageira ----------- **Hood, Hood Components, Hood Component, Tailgate and Midgate**
- porta ----------------------- **Doors, Door Components, Door Handle**
- painel dianteiro/traseiro  -- **Rear Body, Body Component, Trunk Panels, Trunk Component, Chassis Components, Quarter Panels, Rocker Panel
Frame Components, Sheet metal components, Front Panel, Filler Panels, Core Support, Body components, Exterior                                  Components, Floor Components, Core support components, Frame, Floor, Side Body, Body panels, Dash Panel, Deck Panel Bed Panels, Floor Component, Floor Pans**  
- guarda-lamas --------------- **Fenders, Quarter Patch**
- para-choques --------------- **Bumper Component, Bumper Panel, Bumper**


Gaskets and Sealing Systems? --- borrachas

**Hood components AND Hood Components etc.** ---> put all in lowercase

In [12]:
# Filtrar pelas categorias escolhidas

categories = [
    'Roof', 'Hood', 'Doors', 'Tailgate and Midgate', 'Trunk Panels', 'Chassis Components',
    'Quarter Panels', 'Rocker Panel Frame Components', 'Sheet metal components', 'Front Panel',
    'Core Support', 'Body panels', 'Dash Panel', 'Deck Panel Bed Panels', 'Fenders', 'Bumper'
]

db_filtered = db[db['Product_type'].isin(categories)]

db_filtered.count()


Name            342
Product_type    342
Composition     332
Finish          312
Weight          342
Length          342
Width           342
Height          342
Models          342
Price           342
dtype: int64

<span style="color:blue">

- there are 342 classic car parts for the categories filtered

### 3.5. Filter feature
<a id='3.5'>

In [13]:
#unique_f = db['Filter'].unique().tolist()
#print(unique_f)

<span style='color:blue'> 


- categories from filter feature are the same as the categories from product type feature.

### 3.6. Composition feature
<a id='3.6'>

In [14]:
unique_composition = db['Composition'].unique().tolist()
print(unique_composition)

['Stamped steel', 'Injection-molded plastic', 'Steel', nan, 'Stainless Steel', 'Die stamped rubber', 'Extruded plastic', 'Metal', 'Die-cut jute', 'EDP-coated', 'Metal/Rubber', 'Die-cast', 'ABS plastic', 'Metal/Plastic', 'Braided steel', 'Rubber', 'Diecast Metal', 'Stamped Steel', 'Chrome or black', 'Plastic/Aluminum', 'Rubber/Steel', 'Plastic', 'Zinc coated', 'Die-cast zinc', 'Magnesium steel spring', 'Injection-molded rubber', 'Case hardened steel', 'Die-cut rubber', 'Bare', 'Fiberglass', 'Spring steel', 'Injection molded plastic', 'EDP-coated black', 'Fiberglass reinforced plastic', 'Aluminum', 'Natural', 'Aluminum/Steel', 'EDP-Coated', 'Stamped from heavy-gauge steel', 'Zinc', 'Chrome-plated']


<span style='color:blue'> 

steel --> Stamped steel, Steel, Stainless Steel, Braided steel, Stamped Steel, Rubber/Steel, Case hardened steel, Spring steel, Aluminium/Steel, Stamped from heavy-gauge steel

Rubber --> Die stamped rubber, Metal/Rubber, Rubber, Rubber/Steel, Injection-molded rubber, Die-cut rubber

Metal --> Metal/Rubber, Metal/Plastic, Diecast Metal

Jute --> Die-cut jute

Die --> Die-cast (index 17 zinc, index 321 rubber), Diecast Metal, Die-cast Zinc, Die-cut rubber

Chrome --> Chrome or black, Chrome-plated

Aluminium --> Plastic/Aluminium, Aluminium, Aluminium/Steel

Zinc --> Zinc coated, Die-cast zinc, Zinc

Bare --> Bare

Fiberglass --> Fiberglass, Fiberglass reinforced plastic

Natural --> Natural

Plastic --> Injection-molded plastic, Extruded plastic, ABS plastic, Metal/Plastic, Plastic/Aluminium, Plastic, Injection molded plastic, EDP-coated, EDP-coated black, Fiberglass reinforced plastic, EDP-Coated

In [15]:
composition_search = db.loc[db['Composition'] == 'Injection-molded rubber']
composition_search


Name  \
90                           GROMMET, FIREWALL, 1968-72 GM, 1 HOLE   
148               GROMMET, FIREWALL, 1968-72 A-BODY/RIV/GP, 3-HOLE   
174  GROMMET, FIREWALL, 1968-72 BUICK/CHEV/PONT/OLDS W/ AC, 2-1/4"   
198    GROMMET, FIREWALL, 1968-72 A-BODY, SINGLE HOLE, 1-3/4" DIA.   
413                       HOOD ADJUST BUMPER, 1978-88 G-BODY, PAIR   

      Product_type              Composition Finish   Weight  Length   Width  \
90   Firewall/Cowl  Injection-molded rubber    NaN  0.1 lbs  1.0 in  1.0 in   
148  Firewall/Cowl  Injection-molded rubber    NaN  0.5 lbs  1.8 in  1.8 in   
174  Firewall/Cowl  Injection-molded rubber    NaN  0.2 lbs  2.5 in  2.5 in   
198  Firewall/Cowl  Injection-molded rubber    NaN  0.1 lbs  2.0 in  2.0 in   
413           Hood  Injection-molded rubber  Black  0.0 lbs  1.0 in  1.0 in   

     Height  \
90   0.2 in   
148  1.0 in   
174  0.5 in   
198  0.5 in   
413  1.0 in   

                                                                                                                                                                                                                                                                                                                                              Models  \
90       ['1968-1972 Buick Riviera', '1968-1972 Buick Skylark', '1968-1972 Chevrolet Chevelle', '1965-1969 Chevrolet Corvair', '1968-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1968-1972 Oldsmobile Cutlass', '1969-1972 Pontiac Grand Prix', '1968-1972 Pontiac GTO', '1968-1972 Pontiac LeMans', '1968-1970 Pontiac Tempest']   
148      ['1968-1972 Buick Riviera', '1968-1972 Buick Skylark', '1968-1972 Chevrolet Chevelle', '1965-1969 Chevrolet Corvair', '1968-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1968-1972 Oldsmobile Cutlass', '1969-1972 Pontiac Grand Prix', '1968-1972 Pontiac GTO', '1968-1972 Pontiac LeMans', '1968-1970 Pontiac Tempest']   
174                                     ['1968-1972 Buick Riviera', '1968-1972 Buick Skylark', '1968-1972 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1968-1972 Oldsmobile Cutlass', '1969-1972 Pontiac Grand Prix', '1968-1972 Pontiac GTO', '1968-1972 Pontiac LeMans', '1968-1970 Pontiac Tempest']   
198                                                                ['1968-1972 Buick Skylark', '1968-1972 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1968-1972 Oldsmobile Cutlass', '1969-1972 Pontiac Grand Prix', '1968-1972 Pontiac GTO', '1968-1972 Pontiac LeMans', '1968-1970 Pontiac Tempest']   
413  ['1982 Buick Grand National', '1984-1987 Buick Grand National', '1978-1987 Buick Regal', '1983-1987 Buick T-Type', '1978-1987 Chevrolet El Camino', '1978-1983 Chevrolet Malibu', '1978-1988 Chevrolet Monte Carlo', '1978-1987 GMC Caballero', '1978-1988 Oldsmobile Cutlass', '1982-1986 Pontiac Bonneville', '1978-1987 Pontiac Grand Prix']   

            Price  
90    MSRP: $9.99  
148  MSRP: $14.49  
174   MSRP: $7.99  
198  MSRP: $17.49  
413   MSRP: $9.99

In [16]:
db.iloc[56]


Name                                                                              SOLENOID, COWL INDUCTION SYSTEM, 1970-72 CHEVELLE/EL CAMINO
Product_type                                                                                                                  Hood Components
Composition                                                                                                                  Plastic/Aluminum
Finish                                                                                                                                    NaN
Weight                                                                                                                                0.4 lbs
Length                                                                                                                                 3.5 in
Width                                                                                                                                  2.7 in
Height

### 3.7. Finish feature
<a id='3.7'>

In [17]:
unique_finish = db['Finish'].unique().tolist()
print(unique_finish)

['Natural', 'Black', 'Zinc coated', nan, 'Polished', 'Included', 'Black Oxide', 'Rubber backed', 'Galvanized', 'EDP coated', 'Chrome-Plated', 'Black oxide', 'Plastic sleeves', 'EDP Coating', 'Black Coating', 'EDP Coated', 'Black Paint', 'Rubber', 'Metal', 'Plastic', 'EDP', 'EDP or Zinc coated', 'Textured', 'Primer finish', 'EDP-coated', 'EDP-coated black', 'Primer', 'Zinc Plated', 'Complete with hardware', 'Black or Chrome', 'EDP-Coated', 'Natural/EDP Coating', 'Weld-through primer', 'Black enamel', 'Chrome/EDP Coating', 'Chrome']


<span style='color:blue'> 

**EDP coated** AND **EDP** AND **EDP Coating** AND **EDP-Coated** AND **EDP Coated** --> **EDP coated** (remove capital letters, maintain only acronyms as EDP and remove hyphens)

**Primer finish** AND **Primer** --> **Primer finish**

In [18]:
finish_search = db.loc[db['Finish'] == 'EDP coated']
finish_search

Name  \
16             HARDWARE, INNER FENDER, 1964-72 CHEV/EL CAM/MONTE/GTO/ LEM/TEMP   
43                                INNER FENDERWELL, 1968-72 CHEVELLE/EL CAMINO   
46                                                  1970-72 MONTE CARLO FENDER   
51                         INNER FENDERWELLS, 1968-72 CHEVELLE/EL CAMINO, PAIR   
55   PATCH PANEL, 68-9 CHEVELLE, 68-72 EL CAM FRONT FENDER (BEHIND WHEEL WELL)   
..                                                                         ...   
793                                DOOR SHELLS, 1978-85 EL CAMINO/MALIBU, PAIR   
796                                      FENDER, 1965 GTO/TEMPEST/LEMANS, PAIR   
798              FLOOR PAN, COMPLETE, 1968-69 EL CAMINO, W/O ROCKERS OR BRACES   
802                                           DOOR SHELLS, 1968 CHEVELLE, PAIR   
803                                    1970-72 MONTE CARLO FRONT FENDERS, PAIR   

                         Product_type    Composition      Finish    Weight  \
16   Hardware, Fasteners and Fittings          Steel  EDP coated   1.4 lbs   
43                            Fenders  Stamped Steel  EDP coated  10.1 lbs   
46                            Fenders          Steel  EDP coated  41.0 lbs   
51                            Fenders  Stamped Steel  EDP coated  20.2 lbs   
55                            Fenders          Steel  EDP coated   1.5 lbs   
..                                ...            ...         ...       ...   
793                             Doors          Steel  EDP coated  80.0 lbs   
796                           Fenders          Steel  EDP coated  26.0 lbs   
798                        Floor Pans          Steel  EDP coated  70.0 lbs   
802                             Doors          Steel  EDP coated  50.0 lbs   
803                           Fenders          Steel  EDP coated  62.0 lbs   

      Length    Width   Height  \
16    4.7 in   3.4 in   1.7 in   
43   45.0 in  21.0 in  21.0 in   
46   70.0 in  29.0 in  17.0 in   
51   45.0 in  21.0 in  21.0 in   
55   15.0 in  11.0 in   5.0 in   
..       ...      ...      ...   
793  61.3 in  30.0 in   9.3 in   
796  62.0 in  29.0 in  14.0 in   
798  61.0 in  55.0 in   8.0 in   
802   0.0 in   0.0 in   0.0 in   
803  65.0 in  60.0 in  14.0 in   

                                                                                                                                                                                     Models  \
16   ['1964-1972 Chevrolet Chevelle', '1964-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1964-1972 Pontiac GTO', '1964-1972 Pontiac LeMans', '1964-1970 Pontiac Tempest']   
43                                                                                                                        ['1968-1972 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino']   
46                                                                                                                                                      ['1970-1972 Chevrolet Monte Carlo']   
51                                                                                                                        ['1968-1972 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino']   
55                                                                                                                        ['1968-1969 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino']   
..                                                                                                                                                                                      ...   
793                                                                                              ['1978-1985 Chevrolet El Camino', '1978-1983 Chevrolet Malibu', '1978-1985 GMC Caballero']   
796                                                                                                                     ['1965 Pontiac GTO', '1965 Pontiac LeMans', '1965 Pontiac Tempest']   
798                     

### 3.8. Dimensions feature
<a id='3.8'>

In [19]:
db['Weight'].describe()

count         804
unique        196
top       4.5 lbs
freq           27
Name: Weight, dtype: object

### 3.9. Model feature
<a id='3.9'>

In [20]:
unique_models = db['Models'].unique().tolist()
len(unique_models)

222

### 3.10. Price feature
<a id='3.10'>

In [21]:
unique_price = db['Price'].unique().tolist()
print(unique_price)

['MSRP: $6.99', 'MSRP: $26.99', 'MSRP: $14.99', 'MSRP: $36.99', 'MSRP: $12.99', 'MSRP: $8.99', 'MSRP: $17.99', 'MSRP: $7.49', 'MSRP: $52.99', 'MSRP: $49.99', 'MSRP: $74.99', 'MSRP: $0.99', 'MSRP: $9.99', 'MSRP: $39.99', 'MSRP: $69.99', 'MSRP: $16.99', 'MSRP: $79.99', 'MSRP: $32.99', 'MSRP: $11.99', 'MSRP: $24.99', 'MSRP: $19.99', 'MSRP: $89.99', 'MSRP: $109.99', 'MSRP: $10.99', 'MSRP: $46.99', 'MSRP: $159.99', 'MSRP: $143.99', 'MSRP: $126.99', 'MSRP: $58.99', 'MSRP: $498.99', 'MSRP: $149.99', 'MSRP: $124.99', 'MSRP: $392.99', 'MSRP: $287.99', 'MSRP: $8.49', 'MSRP: $29.99', 'MSRP: $59.99', 'MSRP: $42.99', 'MSRP: $13.99', 'MSRP: $65.99', 'MSRP: $5.39', 'MSRP: $332.99', 'MSRP: $13.69', 'MSRP: $64.99', 'MSRP: $48.99', 'MSRP: $86.99', 'MSRP: $117.99', 'MSRP: $62.99', 'MSRP: $479.99', 'MSRP: $3.99', 'MSRP: $119.99', 'MSRP: $299.99', 'MSRP: $45.99', 'MSRP: $432.99', 'MSRP: $428.99', 'MSRP: $449.99', 'MSRP: $399.99', 'MSRP: $199.99', 'MSRP: $178.99', 'MSRP: $129.99', 'MSRP: $112.99', 'MSRP: $4

## 4. Data cleaning
<a id='4'>

### 4.1. Standardization of formats
<a id='4.1'>

In [22]:
# All lowercase

db['Product_type'] = db['Product_type'].str.lower()
#db['Filter'] = db['Filter'].str.lower()
db['Composition'] = db['Composition'].str.lower()
db['Finish'] = db['Finish'].str.lower()

In [23]:
# Unique format for finish feature

#EDP coated AND EDP AND EDP Coating AND EDP-Coated AND EDP Coated --> EDP coated (remove capital letters, maintain only acronyms as EDP)

words = ['edp coated', 'edp', 'edp coating', 'edp-coated']

db['Finish'] = db['Finish'].apply(lambda x: 'EDP coated' if x in words else x)
        

#Primer finish AND Primer --> Primer finish

db['Finish'] = db['Finish'].apply(lambda x:'primer finish' if x == 'primer' else x)


In [24]:
unique_finish = db['Finish'].unique().tolist()
print(unique_finish)

['natural', 'black', 'zinc coated', nan, 'polished', 'included', 'black oxide', 'rubber backed', 'galvanized', 'EDP coated', 'chrome-plated', 'plastic sleeves', 'black coating', 'black paint', 'rubber', 'metal', 'plastic', 'edp or zinc coated', 'textured', 'primer finish', 'edp-coated black', 'zinc plated', 'complete with hardware', 'black or chrome', 'natural/edp coating', 'weld-through primer', 'black enamel', 'chrome/edp coating', 'chrome']


<span style='color:blue'> 

Rubber/Steel --> Composition 'Steel'

Aluminum/Steel --> Composition 'Aluminum' (index 618 and 716), Composition 'Steel' (index 331 and 583)

Metal/Rubber --> Composition 'Metal, Rubber' para brisas tem ambos

Metal/Plastic --> Composition 'Metal, Plastic' sistema induçao de ar tem ambos

Chrome or black --> remover index 47, it's a motor component, not a body component

Plastic/Aluminum --> Composition 'Plastic, Aluminium' sistema induçao do capô

In [25]:
# alter base on indices
db.loc[[331,583], 'Composition'] = 'steel production chromium steel 18/8 hot rolled'
db.loc[[307,355], 'Composition'] = 'steel production chromium steel 18/8 hot rolled'
db.loc[[686], 'Composition'] = 'steel production converted' # bracket, hard steel
db.loc[[618,716], 'Composition'] = 'sheet rolling aluminium'
db.loc[[321], 'Composition'] = 'rubber production' # as rubber is the most common for gaskets
db.loc[[17], 'Composition'] = 'zinc production' # as zinc is the most common for emblems in hoods
db.loc[[65], 'Composition'] = 'polyamide' # most common for Flapper, Valve Frame
db.loc[[93], 'Composition'] = 'polypropylene' # most common for Inner Fenderwells
db.loc[[135], 'Composition'] = 'polyurethane production' # most common for Impact Strip
db.loc[[341], 'Composition'] = 'acrylonitrile-butadiene-styrene' # vent ducts
db.loc[[25], 'Composition'] = 'polyamide' 
db.loc[[56], 'Composition'] = 'polyamide, sheet rolling aluminium' 
db.loc[[208], 'Finish'] = 'electrostatic paint'
db.loc[[421], 'Finish'] = 'chromium coat electroplating'
db.loc[[601,626,638,759], 'Finish'] = 'chromium coat electroplating'
db.loc[[545,596,644,735], 'Finish'] = 'acrylonitrile-butadiene-styrene'


In [26]:
#alter based on words of composition the finish column
db.loc[db['Composition'] == 'edp-coated', 'Finish'] = 'electrostatic paint'
db.loc[db['Composition'] == 'chrome or black', 'Finish'] = 'black chrome'
db.loc[db['Composition'] == 'zinc coated', 'Finish'] = 'zinc coating pieces'
db.loc[db['Composition'] == 'magnesium steel spring', 'Finish'] = 'zinc coating pieces'
db.loc[db['Composition'] == 'chrome-plated', 'Finish'] = 'chromium coat electroplating'

# Alter based on words
db['Composition'] = db['Composition'].replace('die-cast zinc', 'zinc production')
db['Composition'] = db['Composition'].replace('die stamped', 'rubber production')
db['Composition'] = db['Composition'].replace('stamped steel', 'steel production chromium steel 18/8 hot rolled') # steel include stamp
db['Composition'] = db['Composition'].replace('rubber/steel', 'steel production chromium steel 18/8 hot rolled')
db['Composition'] = db['Composition'].replace('rubber', 'rubber production')
db['Composition'] = db['Composition'].replace('aluminum', 'sheet rolling aluminium')
db['Composition'] = db['Composition'].replace('zinc', 'zinc production')
db['Composition'] = db['Composition'].replace('steel', 'steel production chromium steel 18/8 hot rolled')
db['Composition'] = db['Composition'].replace('braided steel', 'reinforcing steel')
db['Composition'] = db['Composition'].replace('metal/rubber', 'steel production chromium steel 18/8 hot rolled, rubber production')
db['Composition'] = db['Composition'].replace('extruded plastic', 'polypropylene') # firewall wire
db['Composition'] = db['Composition'].replace('edp-coated black', 'steel production chromium steel 18/8 hot rolled')
db['Composition'] = db['Composition'].replace('die stamped rubber', 'rubber production')
db['Composition'] = db['Composition'].replace('metal', 'steel production chromium steel 18/8 hot rolled') # as this is the most general metal used in classic car body parts
db['Composition'] = db['Composition'].replace('die-cut jute', 'jute production')
db['Composition'] = db['Composition'].replace('edp-coated', 'steel production chromium steel 18/8 hot rolled') # as this is the most general metal used in classic car body parts and didnt have any material for hood hinge (dobradiça capo)
db['Composition'] = db['Composition'].replace('abs plastic', 'acrylonitrile-butadiene-styrene')
db['Composition'] = db['Composition'].replace('diecast metal', 'steel production chromium steel 18/8 hot rolled')
db['Composition'] = db['Composition'].replace('chrome or black', 'sheet rolling aluminium') # classic car air cleaner most general
db['Composition'] = db['Composition'].replace('zinc coated', 'steel production chromium steel 18/8 hot rolled') # hood catch (tranca capo)
db['Composition'] = db['Composition'].replace('magnesium steel spring', 'steel production chromium steel 18/8 hot rolled') # springs(molas)
db['Composition'] = db['Composition'].replace('case hardened steel', 'steel production converted low-alloyed')
db['Composition'] = db['Composition'].replace('die-cut rubber', 'rubber production') #gaskets
db['Composition'] = db['Composition'].replace('bare', 'steel production chromium steel 18/8 hot rolled')#  weatherstrip channel (aloja borracha do capo)
db['Composition'] = db['Composition'].replace('spring steel', 'steel production converted') # hard steel
db['Composition'] = db['Composition'].replace('injection-molded plastic', 'polypropylene') # Cover, Washer Pump
db['Composition'] = db['Composition'].replace('injection-molded plastic', 'acrylonitrile-butadiene-styrene') # screen, cowl vent
db['Composition'] = db['Composition'].replace('stamped from heavy-gauge steel', 'steel production converted') # hard steel and SHIFTER HUMP
db['Composition'] = db['Composition'].replace('die-cast zinc', 'zinc production')
db['Composition'] = db['Composition'].replace('die stamped rubber', 'rubber production')
db['Composition'] = db['Composition'].replace('fiberglass', 'glass fibre reinforced plastic')
db['Composition'] = db['Composition'].replace('injection molded plastic', 'polypropylene')
db['Composition'] = db['Composition'].replace('stainless steel', 'steel production chromium steel 18/8 hot rolled')
db['Composition'] = db['Composition'].replace('fiberglass reinforced plastic', 'glass fibre reinforced plastic')
db['Composition'] = db['Composition'].replace('injection-molded rubber', 'rubber production')


In [27]:
db['Finish'] = db['Finish'].replace('black', 'electrostatic paint')
db['Finish'] = db['Finish'].replace('black oxide', 'black chrome') # the most indentic
db['Finish'] = db['Finish'].replace('polished', np.nan)  # because polished does not involved energy consuming processes, as it is a manual one
db['Finish'] = db['Finish'].replace('included', np.nan) # because it does not imply a process
db['Finish'] = db['Finish'].replace('natural', np.nan) # because it does not imply a process
db['Finish'] = db['Finish'].replace('rubber backed', 'seal production natural rubber')
db['Finish'] = db['Finish'].replace('galvanized', 'zinc coating pieces')
db['Finish'] = db['Finish'].replace('EDP coated', 'powder coating steel') # the most similar to EDP coat, as it is a coat over steel
db['Finish'] = db['Finish'].replace('chrome-plated', 'chromium coat electroplating')
db['Finish'] = db['Finish'].replace('plastic sleeves', 'plastic film')
db['Finish'] = db['Finish'].replace('black coating', 'electrostatic paint')
db['Finish'] = db['Finish'].replace('black paint', 'electrostatic paint')
db['Finish'] = db['Finish'].replace('edp or zinc coated', 'zinc coating pieces') #because raimundo did not mentioned edp
db['Finish'] = db['Finish'].replace('textured', np.nan) # because probably does not imply energy methods
db['Finish'] = db['Finish'].replace('primer finish', 'electrostatic paint')
db['Finish'] = db['Finish'].replace('edp-coated black', 'selective coating stainless steel black chrome')
db['Finish'] = db['Finish'].replace('primer', 'electrostatic paint')
db['Finish'] = db['Finish'].replace('zinc plated', 'zinc coating pieces')
db['Finish'] = db['Finish'].replace('metal', 'zinc coating pieces') # as it is the most common for RETAINER, TRUNK LOCK (suporte bagagem)
db['Finish'] = db['Finish'].replace('complete with hardware', np.nan) # does not imply energy
db['Finish'] = db['Finish'].replace('natural/edp coating', 'powder coating steel')  # as natural does not imply energy
db['Finish'] = db['Finish'].replace('weld-through primer', 'electrostatic paint')
db['Finish'] = db['Finish'].replace('black enamel', 'enamelling')
db['Finish'] = db['Finish'].replace('chrome', 'chromium coat electroplating')
db['Finish'] = db['Finish'].replace('zinc coated', 'zinc coating pieces')
db['Finish'] = db['Finish'].replace('plastic', 'electrostatic paint') # as this piece is described as to have black painting in his name
db['Finish'] = db['Finish'].replace('rubber', 'seal production natural rubber')


In [28]:
composition_unique = db['Composition'].unique().tolist()
print(composition_unique)

['steel production chromium steel 18/8 hot rolled', 'polypropylene', nan, 'rubber production', 'jute production', 'steel production chromium steel 18/8 hot rolled, rubber production', 'zinc production', 'acrylonitrile-butadiene-styrene', 'polyamide', 'reinforcing steel', 'sheet rolling aluminium', 'polyamide, sheet rolling aluminium', 'steel production converted low-alloyed', 'polyurethane production', 'glass fibre reinforced plastic', 'steel production converted']


In [29]:
finish_unique = db['Finish'].unique().tolist()
print(finish_unique)

[nan, 'electrostatic paint', 'zinc coating pieces', 'black chrome', 'seal production natural rubber', 'powder coating steel', 'chromium coat electroplating', 'plastic film', 'selective coating stainless steel black chrome', 'enamelling', 'acrylonitrile-butadiene-styrene']


### 4.2. Noise removal
<a id='4.2'>

In [30]:
#remove lbs and in

db['Weight'] = db['Weight'].str.replace(' lbs', '')
db['Length'] = db['Length'].str.replace(' in', '')
db['Width'] = db['Width'].str.replace(' in', '')
db['Height'] = db['Height'].str.replace(' in', '')

In [31]:
# remove MSRP: $

db['Price'] = db['Price'].str.replace(r'^MSRP: \$', '', regex=True)

In [32]:
db.head()

Name  \
0                                DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)   
1                             SCREEN, COWL VENT, 1968-72 A-BODY, 1969-72 GRAND PRIX, W/ AC   
2                                               SPRING, HOOD HINGE, 1964-67/1970-72 A-BODY   
3  1964-1977 HOOD ADJUSTER BOLT KIT, GM A-BODY, RIVIERA, 1959-1963 BONNEVILLE AND CATALINA   
4                                             HOOD PIN SET, 1970-72 CHEVELLE/EL CAMINO, SS   

                       Product_type  \
0  hardware, fasteners and fittings   
1         cowl induction components   
2                              hood   
3                     bolts & studs   
4                   hood components   

                                       Composition               Finish  \
0  steel production chromium steel 18/8 hot rolled                  NaN   
1                                    polypropylene  electrostatic paint   
2  steel production chromium steel 18/8 hot rolled  zinc coating pieces   
3                                              NaN                  NaN   
4  steel production chromium steel 18/8 hot rolled                  NaN   

  Weight Length Width Height  \
0    0.2    5.0   3.5    0.2   
1    0.3   14.8   4.8    1.5   
2    2.0    7.4   1.8    1.2   
3    0.1    2.0   1.2    1.0   
4    3.0    5.6   3.4    1.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Models  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      ['1967-1976 Buick Riviera', '1967-1972 Buick Skylark', '1967-1977 Cadillac DeVille', '1967-1977 Cadillac Eldorado', '1967-1977 Cadillac Fle

### 4.3. Data type conversion
<a id='4.3'>

In [33]:
# Object type to numeric type

#Weight, Length, Width, Height

db['Weight'] = db['Weight'].astype(float)
db['Length'] = db['Length'].astype(float)
db['Width'] = db['Width'].astype(float)
db['Height'] = db['Height'].astype(float)
db['Price'] = db['Price'].str.replace(',', '').astype(float)

In [34]:
db['Name'] = db['Name'].astype(str)
db['Product_type'] = db['Product_type'].astype(str)
#db['Filter'] = db['Filter'].astype(str)


In [35]:
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 804 entries, 0 to 803
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          804 non-null    object 
 1   Product_type  804 non-null    object 
 2   Composition   746 non-null    object 
 3   Finish        671 non-null    object 
 4   Weight        804 non-null    float64
 5   Length        804 non-null    float64
 6   Width         804 non-null    float64
 7   Height        804 non-null    float64
 8   Models        804 non-null    object 
 9   Price         804 non-null    float64
dtypes: float64(5), object(5)
memory usage: 62.9+ KB


## 5. Data transformation
<a id='5'>

In [36]:
# USD to EUR
convertion_tax_USD_EUR = 0.85
db['Price'] = db['Price'] * convertion_tax_USD_EUR


# in to mm
convertion_tax_in_mm = 25.4
db['Length'] = db['Length'] * convertion_tax_in_mm
db['Width'] = db['Width'] * convertion_tax_in_mm
db['Height'] = db['Height'] * convertion_tax_in_mm


# lb to kg
convertion_tax_lb_kg = 0.453592
db['Weight'] = db['Weight'] * convertion_tax_lb_kg


# round to 2 decimal units
db['Weight'] = db['Weight'].round(2)
db['Length'] = db['Length'].round(2)
db['Width'] = db['Width'].round(2)
db['Height'] = db['Height'].round(2)
db['Price'] = db['Price'].round(2)


In [37]:
db = db.rename(columns={'Price': 'Price (€)'})
db = db.rename(columns={'Weight': 'Weight (kg)'})
db = db.rename(columns={'Length': 'Length (mm)'})
db = db.rename(columns={'Width': 'Width (mm)'})
db = db.rename(columns={'Height': 'Height (mm)'})

In [38]:
db.head(2)

Name  \
0     DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)   
1  SCREEN, COWL VENT, 1968-72 A-BODY, 1969-72 GRAND PRIX, W/ AC   

                       Product_type  \
0  hardware, fasteners and fittings   
1         cowl induction components   

                                       Composition               Finish  \
0  steel production chromium steel 18/8 hot rolled                  NaN   
1                                    polypropylene  electrostatic paint   

   Weight (kg)  Length (mm)  Width (mm)  Height (mm)  \
0         0.09       127.00       88.90         5.08   
1         0.14       375.92      121.92        38.10   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Models  \
0  ['1967-1976 Buick Riviera', '1967-1972 Buick Skylark', '1967-1977 Cadillac DeVille', '1967-1977 Cadillac Eldorado', '1967-1977 Cadillac Fleetwood', '1967-1981 Chevrolet Camaro', '1967-1977 Chevrolet Chevelle', '1967-1977 Chevrolet El Camino', '1964-1972 Chevrolet Impala', '1970-1977 Chevrolet Monte Carlo', '1962-1972 Chevrolet Nova', '1971-1972 GMC Sprint', '1967-1977 Oldsmobile Cutlass', '1967-1977 Pontiac Bonneville', '1967-1977 Pontiac Catalina', '1967-1981 Pontiac Firebird', '1967-1977 Pontiac Grand Prix', '1967-1973 Pontiac GTO', '1967-1973 Pontiac LeMans', '1967-1970 Pontiac Tempest']   
1                                                                                                                                                                                                                                                                                                                                      ['1968-1972 Buick Skylark', '1968-1972 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1968-1972 Oldsmobile Cutlass', '1969-1972 Pontiac Grand Prix', '1968-1972 Pontiac GTO', '1968-1972 Pontiac LeMans', '1968-1970 Pontiac Tempest']   

   Price (€)  
0       5.94  
1      22.94

In [39]:
db.describe()

Weight (kg)  Length (mm)   Width (mm)  Height (mm)   Price (€)
count   804.000000   804.000000   804.000000   804.000000  804.000000
mean      8.745560   960.312711   448.843905   153.414104  187.245983
std      11.131222   647.179631   389.646174   127.593877  174.298884
min       0.000000     0.000000     0.000000     0.000000    0.840000
25%       1.270000   424.815000   139.700000    50.800000   59.490000
50%       3.450000   952.500000   330.200000   127.000000  124.090000
75%      13.060000  1524.000000   670.560000   228.600000  267.102500
max      63.500000  2540.000000  1778.000000   800.100000  934.990000

## 6. Data integration
<a id='6'>

In [40]:
# add unique ID
db['ID_part'] = db.index 

### 6.1. Models table

<a id='6.1'>

In [41]:
db.head(3)

Name  \
0     DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)   
1  SCREEN, COWL VENT, 1968-72 A-BODY, 1969-72 GRAND PRIX, W/ AC   
2                    SPRING, HOOD HINGE, 1964-67/1970-72 A-BODY   

                       Product_type  \
0  hardware, fasteners and fittings   
1         cowl induction components   
2                              hood   

                                       Composition               Finish  \
0  steel production chromium steel 18/8 hot rolled                  NaN   
1                                    polypropylene  electrostatic paint   
2  steel production chromium steel 18/8 hot rolled  zinc coating pieces   

   Weight (kg)  Length (mm)  Width (mm)  Height (mm)  \
0         0.09       127.00       88.90         5.08   
1         0.14       375.92      121.92        38.10   
2         0.91       187.96       45.72        30.48   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Models  \
0  ['1967-1976 Buick Riviera', '1967-1972 Buick Skylark', '1967-1977 Cadillac DeVille', '1967-1977 Cadillac Eldorado', '1967-1977 Cadillac Fleetwood', '1967-1981 Chevrolet Camaro', '1967-1977 Chevrolet Chevelle', '1967-1977 Chevrolet El Camino', '1964-1972 Chevrolet Impala', '1970-1977 Chevrolet Monte Carlo', '1962-1972 Chevrolet Nova', '1971-1972 GMC Sprint', '1967-1977 Oldsmobile Cutlass', '1967-1977 Pontiac Bonneville', '1967-1977 Pontiac Catalina', '1967-1981 Pontiac Firebird', '1967-1977 Pontiac Grand Prix', '1967-1973 Pontiac GTO', '1967-1973 Pontiac LeMans', '1967-1970 Pontiac Tempest']   
1                                                                                                                                                                                                                                                                                                                                      ['1968-1972 Buick Skylark', '1968-1972 Chevrolet Chevelle', '1968-1972 Chevrolet El Camino', '1970-1972 Chevrolet Monte Carlo', '1968-1972 Oldsmobile Cutlass', '1969-1972 Pontiac Grand Prix', '1968-1972 Pontiac GTO', '1968-1972 Pontiac LeMans', '1968-1970 Pontiac Tempest']   
2                                                                     ['1963-1966 Chevrolet Bel Air', '1963-1966 Chevrolet Biscayne', '1967-1969 Chevrolet Camaro', '1966 Chevrolet Caprice', '1964-1967 Chevrolet Chevelle', '1970-1972 Chevrolet Chevelle', '1964-1967 Chevrolet El Camino', '1970-1972 Chevrolet El Camino', '1970 Chevrolet Impala', '1970-1972 Chevrolet Monte Carlo', '1971-1972 GMC Sprint', '1964-1967 Oldsmobile Cutlass', '1967-1969 Pontiac Firebird', '1964-1967 Pontiac GTO', '1971-1972 Pontiac GTO', '1964-1967 Pontiac LeMans', '1971-1972 Pontiac LeMans', '1964-1967 Pontiac Tempest']   

   Price (€)  ID_part  
0       5.94        0  
1      22.94        1  
2      12.74        2

### ID Model 

In [42]:
lists = db['Models']
import ast

# str to list
list_data = [ast.literal_eval(d) for d in lists]

# unique list
flat_list = [item for sublist in list_data for item in sublist]

#unique values
unique_values = set(flat_list)

# new list
unique_values_list = list(unique_values)


len(unique_values_list)


445

In [43]:
# convert str column to list
db['Models'] = db['Models'].apply(ast.literal_eval)

In [44]:
# map models IDs to new column ID model
id_model, name_model = pd.factorize(unique_values_list)


model_to_id = dict(zip(name_model, id_model))


def map_models_to_ids(models):
    return [model_to_id.get(model) for model in models]

# Aplicando a função para criar a nova coluna 'ID_model'
db['ID_model'] = db['Models'].apply(map_models_to_ids)

C:\Users\InesC\AppData\Local\Temp\ipykernel_14224\3311165932.py:2: FutureWarning: factorize with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.
  id_model, name_model = pd.factorize(unique_values_list)


### ID product type

In [45]:
unique_products = db['Product_type'].unique()

id_product_type, name_product_type = pd.factorize(unique_products)

products_to_id = dict(zip(name_product_type, id_product_type))

def map_products_to_ids(product):
    return products_to_id.get(product)

db['ID_product_type'] = db['Product_type'].apply(map_products_to_ids)

### ID composition

In [46]:
split_compositions = db['Composition'].dropna().str.split(',').explode().str.strip()

unique_compositions = split_compositions.unique()
id_composition, name_composition = pd.factorize(unique_compositions)
composition_to_id = dict(zip(name_composition, id_composition))

print(composition_to_id)

def map_compositions_to_ids(composition):
    if pd.isna(composition):
        return [-1]  # Retorna uma lista com -1 para NaN
    compositions = [comp.strip() for comp in composition.split(',')]
    ids = [composition_to_id.get(comp, -1) for comp in compositions]
    return ids 

# Aplicar a função ao DataFrame
db['ID_composition'] = db['Composition'].apply(map_compositions_to_ids)

{'steel production chromium steel 18/8 hot rolled': 0, 'polypropylene': 1, 'rubber production': 2, 'jute production': 3, 'zinc production': 4, 'acrylonitrile-butadiene-styrene': 5, 'polyamide': 6, 'reinforcing steel': 7, 'sheet rolling aluminium': 8, 'steel production converted low-alloyed': 9, 'polyurethane production': 10, 'glass fibre reinforced plastic': 11, 'steel production converted': 12}


In [47]:
#unique_ids = db['ID_composition'].unique()
#print(unique_ids)

### ID finish

In [48]:
all_finishes = db['Finish'].dropna().str.split(',').explode().str.strip()
unique_finishes = all_finishes.unique()
print(unique_finishes)

id_finish, name_finish = pd.factorize(unique_finishes)
#id_finish = id_finish.astype(int)

finishes_to_id = dict(zip(name_finish, id_finish))
#print(finishes_to_id)

def map_finishes_to_ids(finish):
    if pd.isna(finish):
        return -1  # Valor para NaN
    finishes = [finish.strip() for finish in finish.split(',')]
    ids = [finishes_to_id.get(finish, -1) for finish in finishes]
    if not ids:
        return -1
    return ids[0]


db['ID_finish'] = db['Finish'].apply(map_finishes_to_ids)

['electrostatic paint' 'zinc coating pieces' 'black chrome'
 'seal production natural rubber' 'powder coating steel'
 'chromium coat electroplating' 'plastic film'
 'selective coating stainless steel black chrome' 'enamelling'
 'acrylonitrile-butadiene-styrene']


In [49]:
unique = db['ID_finish'].unique()
print(unique)

[-1  0  1  2  3  4  5  6  7  8  9]


In [50]:
#db['ID_composition'] = db['ID_composition'].astype(int)
#db['ID_finish'] = db['ID_finish'].astype(int)

In [51]:
db[db['Composition'] == 'steel production chromium steel 18/8 hot rolled, rubber production']

Name      Product_type  \
15         STOP CLIP, WINDSHIELD GLASS, 1968-72 A-BODY  window component   
436  STOP CLIP, WINDSHIELD GLASS, 1968-72 A-BODY, PAIR  window component   

                                                            Composition  \
15   steel production chromium steel 18/8 hot rolled, rubber production   
436  steel production chromium steel 18/8 hot rolled, rubber production   

                  Finish  Weight (kg)  Length (mm)  Width (mm)  Height (mm)  \
15   zinc coating pieces         0.00        91.44       58.42         7.62   
436  zinc coating pieces         0.05        91.44       58.42         7.62   

                                                                                                                                                                                                                                                                                                                    Models  \
15   [1968-1972 Buick Skylark, 1967-1969 Chevrolet Camaro, 1968-1972 Chevrolet Chevelle, 1968-1972 Chevrolet El Camino, 1970-1972 Chevrolet Monte Carlo, 1967-1979 Chevrolet Nova, 1968-1972 Oldsmobile Cutlass, 1969-1972 Pontiac Grand Prix, 1968-1972 Pontiac GTO, 1968-1972 Pontiac LeMans, 1968-1970 Pontiac Tempest]   
436  [1968-1972 Buick Skylark, 1967-1969 Chevrolet Camaro, 1968-1972 Chevrolet Chevelle, 1968-1972 Chevrolet El Camino, 1970-1972 Chevrolet Monte Carlo, 1967-1979 Chevrolet Nova, 1968-1972 Oldsmobile Cutlass, 1969-1972 Pontiac Grand Prix, 1968-1972 Pontiac GTO, 1968-1972 Pontiac LeMans, 1968-1970 Pontiac Tempest]   

     Price (€)  ID_part                                              ID_model  \
15        8.49       15  [75, 379, 373, 175, 400, 84, 306, 252, 299, 39, 304]   
436      16.99      436  [75, 379, 373, 175, 400, 84, 306, 252, 299, 39, 304]   

     ID_product_type ID_composition  ID_finish  
15                 9         [0, 2]          1  
436                9         [0, 2]          1

In [52]:
def split_model_info(model_info):
    parts = model_info.split(' ')
    year_range = parts[0].split('-')
    ano_inicio = year_range[0]
    ano_fim = year_range[1] if len(year_range) > 1 else year_range[0]
    marca = parts[1]
    modelo = ' '.join(parts[2:])
   
    
    # Concatenar todas as partes em uma única string
    return f"{marca}, {modelo}, {ano_inicio}, {ano_fim}"

# Aplicar a função a todos os valores
split_data = [split_model_info(model) for model in flat_list]

# Criar um novo DataFrame com a coluna única
db22 = pd.DataFrame(split_data, columns=['Model_Info'])


In [53]:
db22.describe()

Model_Info
count                                 2714
unique                                 445
top     Chevrolet, Monte Carlo, 1970, 1972
freq                                   111

In [54]:
def split_model_info(model_info):
    parts = model_info.split(' ')
    year_range = parts[0].split('-')
    ano_inicio = year_range[0]
    ano_fim = year_range[1] if len(year_range) > 1 else year_range[0]
    marca = parts[1]
    modelo = ' '.join(parts[2:])
    ID_model = model_to_id.get(model_info)
    return [ID_model, marca, modelo, ano_inicio, ano_fim]

# Aplicar a função a todos os valores únicos
split_data = [split_model_info(model) for model in unique_values_list]

# Criar um novo DataFrame com as colunas divididas
columns = ['ID_model','Brand', 'Model', 'Start_year', 'End_year']
db2 = pd.DataFrame(split_data, columns=columns)


In [55]:
db2.tail(6)

ID_model      Brand     Model Start_year End_year
439       439  Chevrolet    Impala       1969     1969
440       440  Chevrolet      Nova       1971     1974
441       441    Pontiac  Catalina       1969     1972
442       442      Buick   Skylark       1961     1972
443       443    Pontiac    LeMans       1967     1973
444       444  Chevrolet  Corvette       1973     1975

In [56]:
unique_brand = db2['Brand'].unique().tolist()
print(unique_brand)

['Pontiac', 'Chevrolet', 'Oldsmobile', 'Cadillac', 'Buick', 'GMC']


In [57]:
unique_model = db2['Model'].unique().tolist()
print(unique_model)

['LeMans', 'Nova', 'Catalina', 'Omega', 'Cutlass', 'Grand Prix', 'Tempest', 'Monte Carlo', 'Astre', '60 Special', 'Caprice', 'GTO', 'Bel Air', 'Chevelle', 'Camaro', 'Firebird', 'Pickup', 'Ventura', 'T-Type', 'Fleetwood', 'Century', 'Seville', 'DeVille', 'Skylark', 'Corvette', 'Caballero', 'Electra', 'El Camino', 'Starfire', 'Riviera', 'Bonneville', 'Parisienne', 'Biscayne', 'Brookwood', 'Malibu', 'Ninety-Eight', 'Series 65', 'Impala', 'Regal', 'Eighty-Eight', 'Eldorado', 'Sunbird', 'Toronado', 'Tri-Five', 'Series 62', 'Grand Am', 'Apollo', 'Acadian', 'Monza', 'Grand National', 'Sprint', 'Estate', 'Wildcat', 'Series 60', 'Corvair', 'Lesabre', 'Roadmaster', 'Blazer', 'Series 70', 'Series 63', 'Vega', 'Centurion', 'Series 61', 'Skyhawk', 'Nomad']


In [58]:
unique_startyear = db2['Start_year'].unique().tolist()
print(unique_startyear)

['1969', '1962', '1959', '1973', '1966', '1965', '1970', '1938', '1964', '1968', '1971', '1967', '1961', '1983', '1974', '1980', '1977', '1982', '1976', '1981', '1963', '1975', '1958', '1997', '1954', '1946', '1978', '1994', '1984', '1949', '2010', '1979', '1941', '1955', '1986', '1987', '1953', '1990', '1939', '1936', '2000', '1940', '1985', '1960', '1947']


In [59]:
db2.count() 

ID_model      445
Brand         445
Model         445
Start_year    445
End_year      445
dtype: int64

In [60]:
db2['ID_model'] = db2['ID_model'].astype('object')

In [61]:
db2.describe(include='all')

ID_model      Brand      Model Start_year End_year
count        445        445        445        445      445
unique       445          6         65         45       43
top            0  Chevrolet  El Camino       1964     1972
freq           1        154         30         55       54

### 6.2. Brand table

<a id='6.2'>

In [62]:
unique_brands = db2['Brand'].unique()

id_brand, name_brand = pd.factorize(unique_brands)

brands_to_id = dict(zip(name_brand, id_brand))

def map_brands_to_ids(brand):
    return brands_to_id.get(brand)

db2['ID_brand'] = db2['Brand'].apply(map_brands_to_ids)

In [63]:
db6 = db2.loc[:,['ID_brand','Brand']]
db6 = db6.drop_duplicates()


In [64]:
db6.head(2)

ID_brand      Brand
0         0    Pontiac
1         1  Chevrolet

### 6.3. Compostiion and finishing tables

<a id='6.3'>

In [65]:
import numpy as np

conditions = [
    db['Composition'].str.contains('steel', case=False, na=False),
    db['Composition'].str.contains('plastic', case=False, na=False),
    db['Composition'].str.contains('zinc', case=False, na=False),
    db['Composition'].str.contains('rubber', case=False, na=False),
    db['Composition'].str.contains('chrome', case=False, na=False),
    db['Composition'].str.contains('metal', case=False, na=False),
    db['Composition'].str.contains('edp', case=False, na=False),
    db['Composition'].str.contains('fiberglass', case=False, na=False),
    db['Composition'].str.contains('aluminum', case=False, na=False),
    db['Composition'].str.contains('natural', case=False, na=False),
]

choices = [
    'steel', 'plastic', 'zinc', 'rubber', 'chrome', 
    'metal', 'EDP', 'fiberglass', 'aluminium', 'natural'
]

#db['Composition_material'] = np.select(conditions, choices, default='')

conditions2 = [
    db['Finish'].str.contains('zinc', case=False, na=False),
    db['Finish'].str.contains('polished', case=False, na=False),
    db['Finish'].str.contains('rubber', case=False, na=False),
    db['Finish'].str.contains('galvanized', case=False, na=False),
    db['Finish'].str.contains('metal', case=False, na=False),
    db['Finish'].str.contains('edp', case=False, na=False),
    db['Finish'].str.contains('chrome', case=False, na=False),
    db['Finish'].str.contains('primer', case=False, na=False),
    db['Finish'].str.contains('natural', case=False, na=False),
]

choices2 = [
    'zinc', 'polished', 'rubber', 'galvanized', 
    'metal', 'EDP', 'chrome', 'primer', 'natural'
]

#db['Finish_type'] = np.select(conditions2, choices2, default='')

#db3 = db[['ID_part','Name','Composition', 'Composition_material', 'Finish', 'Finish_type']]

#db3.head()

### 6.3 Product type table
<a id='6.3'>

In [66]:
# Extração dos produtos únicos e suas IDs
unique_products = db['Product_type'].unique()
id_product_type, name_product_type = pd.factorize(unique_products)

products_to_id = dict(zip(name_product_type, id_product_type))

def map_products_to_ids(product):
    return products_to_id.get(product)

db['ID_product_type'] = db['Product_type'].apply(map_products_to_ids)

db5= db[['ID_product_type', 'Product_type']].drop_duplicates()

db5.reset_index(drop=True, inplace=True)


In [67]:
db5.head(2)

ID_product_type                      Product_type
0                0  hardware, fasteners and fittings
1                1         cowl induction components

In [68]:
db5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ID_product_type  77 non-null     int64 
 1   Product_type     77 non-null     object
dtypes: int64(1), object(1)
memory usage: 1.3+ KB


### 6.4 Composition table
<a id='6.4'>

In [69]:
data = [(id, comp) for comp, id in composition_to_id.items()]

# Criar o DataFrame
db7 = pd.DataFrame(data, columns=['ID_composition', 'Composition'])

# Remover duplicatas e redefinir o índice
db7 = db7.drop_duplicates()
db7.reset_index(drop=True, inplace=True)


In [70]:
db7.head(13)

ID_composition                                      Composition
0                0  steel production chromium steel 18/8 hot rolled
1                1                                    polypropylene
2                2                                rubber production
3                3                                  jute production
4                4                                  zinc production
5                5                  acrylonitrile-butadiene-styrene
6                6                                        polyamide
7                7                                reinforcing steel
8                8                          sheet rolling aluminium
9                9           steel production converted low-alloyed
10              10                          polyurethane production
11              11                   glass fibre reinforced plastic
12              12                       steel production converted

In [71]:
db7.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13 entries, 0 to 12
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ID_composition  13 non-null     int64 
 1   Composition     13 non-null     object
dtypes: int64(1), object(1)
memory usage: 340.0+ bytes


### 6.5 Finish table
<a id='6.5'>

In [72]:
data = [(id, finish) for finish, id in finishes_to_id.items()]

# Criar o DataFrame
db8 = pd.DataFrame(data, columns=['ID_finish', 'Finish'])

# Remover duplicatas e redefinir o índice
db8 = db8.drop_duplicates()
db8 = db8.dropna()
db8.reset_index(drop=True, inplace=True)

In [73]:
db8.head(16)

ID_finish                                          Finish
0          0                             electrostatic paint
1          1                             zinc coating pieces
2          2                                    black chrome
3          3                  seal production natural rubber
4          4                            powder coating steel
5          5                    chromium coat electroplating
6          6                                    plastic film
7          7  selective coating stainless steel black chrome
8          8                                      enamelling
9          9                 acrylonitrile-butadiene-styrene

In [74]:
db8.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   ID_finish  10 non-null     int64 
 1   Finish     10 non-null     object
dtypes: int64(1), object(1)
memory usage: 292.0+ bytes


## 7. Data selection
<a id='7'>

In [75]:
# Remove column 'Filter' as it is almost the same as product type

db = db.drop(columns=['Composition', 'Finish', 'Models', 'Product_type'])

db2 = db2.drop(columns=['Brand'])

In [76]:
db2.tail(10)

ID_model     Model Start_year End_year  ID_brand
435      435    Impala       1962     1967         1
436      436   Tempest       1965     1967         0
437      437      Nova       1966     1967         1
438      438      Nova       1985     1988         1
439      439    Impala       1969     1969         1
440      440      Nova       1971     1974         1
441      441  Catalina       1969     1972         0
442      442   Skylark       1961     1972         4
443      443    LeMans       1967     1973         0
444      444  Corvette       1973     1975         1

In [77]:
db.head(1)

Name  Weight (kg)  \
0  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   

   Length (mm)  Width (mm)  Height (mm)  Price (€)  ID_part  \
0        127.0        88.9         5.08       5.94        0   

                                                                                               ID_model  \
0  [164, 174, 131, 390, 147, 295, 153, 429, 251, 178, 161, 416, 125, 225, 378, 409, 113, 195, 443, 210]   

   ID_product_type ID_composition  ID_finish  
0                0            [0]         -1

In [78]:
# split ID model
db4 = db.explode('ID_model').reset_index(drop=True)
#db4.rename(columns={'ID_model':'ID_car'}, inplace=True)


In [79]:
db4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2714 entries, 0 to 2713
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             2714 non-null   object 
 1   Weight (kg)      2714 non-null   float64
 2   Length (mm)      2714 non-null   float64
 3   Width (mm)       2714 non-null   float64
 4   Height (mm)      2714 non-null   float64
 5   Price (€)        2714 non-null   float64
 6   ID_part          2714 non-null   int64  
 7   ID_model         2714 non-null   object 
 8   ID_product_type  2714 non-null   int64  
 9   ID_composition   2714 non-null   object 
 10  ID_finish        2714 non-null   int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 233.4+ KB


In [80]:
db9 = db4.explode('ID_composition').reset_index(drop=True)

In [81]:
# add unique ID
db9['ID'] = db9.index 

In [82]:
db9.head()

Name  Weight (kg)  \
0  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   
1  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   
2  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   
3  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   
4  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   

   Length (mm)  Width (mm)  Height (mm)  Price (€)  ID_part ID_model  \
0        127.0        88.9         5.08       5.94        0      164   
1        127.0        88.9         5.08       5.94        0      174   
2        127.0        88.9         5.08       5.94        0      131   
3        127.0        88.9         5.08       5.94        0      390   
4        127.0        88.9         5.08       5.94        0      147   

   ID_product_type ID_composition  ID_finish  ID  
0                0              0         -1   0  
1                0              0         -1   1  
2                0              0         -1   2  
3                0              0         -1   3  
4                0              0         -1   4

In [83]:
# alter order of columns db

db4 = db4.reindex(columns=['ID','ID_part','ID_model','ID_product_type', 'ID_composition', 'ID_finish', 'Name', 'Weight (kg)', 'Length (mm)', 'Width (mm)', 'Height (mm)', 'Price (€)'])

In [84]:
db4.tail(2)

ID  ID_part ID_model  ID_product_type ID_composition  ID_finish  \
2712 NaN      802      221               33            [0]          4   
2713 NaN      803      400                8            [0]          4   

                                         Name  Weight (kg)  Length (mm)  \
2712         DOOR SHELLS, 1968 CHEVELLE, PAIR        22.68          0.0   
2713  1970-72 MONTE CARLO FRONT FENDERS, PAIR        28.12       1651.0   

      Width (mm)  Height (mm)  Price (€)  
2712         0.0          0.0     620.49  
2713      1524.0        355.6     848.29

In [85]:
db2 = db2.reindex(columns=['ID_model','ID_brand','Model','Start_year', 'End_year'])

In [86]:
db2.head(6)

ID_model  ID_brand       Model Start_year End_year
0        0         0      LeMans       1969     1969
1        1         1        Nova       1962     1979
2        2         0    Catalina       1959     1977
3        3         2       Omega       1973     1977
4        4         2     Cutlass       1966     1967
5        5         0  Grand Prix       1962     1963

In [87]:
#db3.head(2)

In [88]:
db5.head(2)

ID_product_type                      Product_type
0                0  hardware, fasteners and fittings
1                1         cowl induction components

In [89]:
db6.head(2)

ID_brand      Brand
0         0    Pontiac
1         1  Chevrolet

In [90]:
db7.head(2)

ID_composition                                      Composition
0               0  steel production chromium steel 18/8 hot rolled
1               1                                    polypropylene

In [91]:
db8.head(2)

ID_finish               Finish
0          0  electrostatic paint
1          1  zinc coating pieces

In [92]:
db9 = db9.reindex(columns=['ID','ID_part','ID_model','ID_product_type', 'ID_composition', 'ID_finish', 'Name', 'Weight (kg)', 'Length (mm)', 'Width (mm)', 'Height (mm)', 'Price (€)'])

In [93]:
db9.head(2)

ID  ID_part ID_model  ID_product_type ID_composition  ID_finish  \
0   0        0      164                0              0         -1   
1   1        0      174                0              0         -1   

                                                        Name  Weight (kg)  \
0  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   
1  DRAIN PLUG, FLOOR & TRUNK PAN, 1967-77 GM (STAMPED STEEL)         0.09   

   Length (mm)  Width (mm)  Height (mm)  Price (€)  
0        127.0        88.9         5.08       5.94  
1        127.0        88.9         5.08       5.94

In [94]:
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 804 entries, 0 to 803
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             804 non-null    object 
 1   Weight (kg)      804 non-null    float64
 2   Length (mm)      804 non-null    float64
 3   Width (mm)       804 non-null    float64
 4   Height (mm)      804 non-null    float64
 5   Price (€)        804 non-null    float64
 6   ID_part          804 non-null    int64  
 7   ID_model         804 non-null    object 
 8   ID_product_type  804 non-null    int64  
 9   ID_composition   804 non-null    object 
 10  ID_finish        804 non-null    int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 69.2+ KB


In [95]:
db2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445 entries, 0 to 444
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   ID_model    445 non-null    object
 1   ID_brand    445 non-null    int64 
 2   Model       445 non-null    object
 3   Start_year  445 non-null    object
 4   End_year    445 non-null    object
dtypes: int64(1), object(4)
memory usage: 17.5+ KB


In [96]:
#db3.info()

In [97]:
db4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2714 entries, 0 to 2713
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               0 non-null      float64
 1   ID_part          2714 non-null   int64  
 2   ID_model         2714 non-null   object 
 3   ID_product_type  2714 non-null   int64  
 4   ID_composition   2714 non-null   object 
 5   ID_finish        2714 non-null   int64  
 6   Name             2714 non-null   object 
 7   Weight (kg)      2714 non-null   float64
 8   Length (mm)      2714 non-null   float64
 9   Width (mm)       2714 non-null   float64
 10  Height (mm)      2714 non-null   float64
 11  Price (€)        2714 non-null   float64
dtypes: float64(6), int64(3), object(3)
memory usage: 254.6+ KB


In [98]:
db5.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ID_product_type  77 non-null     int64 
 1   Product_type     77 non-null     object
dtypes: int64(1), object(1)
memory usage: 1.3+ KB


In [99]:
db6.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6 entries, 0 to 40
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   ID_brand  6 non-null      int64 
 1   Brand     6 non-null      object
dtypes: int64(1), object(1)
memory usage: 144.0+ bytes


In [100]:
db7.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13 entries, 0 to 12
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ID_composition  13 non-null     int64 
 1   Composition     13 non-null     object
dtypes: int64(1), object(1)
memory usage: 340.0+ bytes


In [101]:
db8.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   ID_finish  10 non-null     int64 
 1   Finish     10 non-null     object
dtypes: int64(1), object(1)
memory usage: 292.0+ bytes


In [102]:
db9.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2740 entries, 0 to 2739
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               2740 non-null   int64  
 1   ID_part          2740 non-null   int64  
 2   ID_model         2740 non-null   object 
 3   ID_product_type  2740 non-null   int64  
 4   ID_composition   2740 non-null   object 
 5   ID_finish        2740 non-null   int64  
 6   Name             2740 non-null   object 
 7   Weight (kg)      2740 non-null   float64
 8   Length (mm)      2740 non-null   float64
 9   Width (mm)       2740 non-null   float64
 10  Height (mm)      2740 non-null   float64
 11  Price (€)        2740 non-null   float64
dtypes: float64(5), int64(4), object(3)
memory usage: 257.0+ KB


## 8. Data exportation
<a id='8'>

In [103]:
db.to_csv('Bodyworkparts_processeddata.csv', sep=';', encoding='utf-8', index=False) # nao usado

In [104]:
db2.to_csv('restoparts_models.csv', sep=';', encoding='utf-8', index=False)
db2.to_csv('restoparts_models_withoutheader.csv', sep=';', encoding='utf-8', index=False, header=False)

In [105]:
#db3.to_csv('restoparts_composition_finish.csv', sep=';', encoding='utf-8', index=False)
#db3.to_csv('restoparts_composition_finish_withoutheader.csv', sep=';', encoding='utf-8', index=False, header=False)

In [106]:
db4.to_csv('restoparts_body_clean_exploded.csv', sep=';', encoding='utf-8', index=False) # original

In [107]:
db5.to_csv('restoparts_product_type.csv', sep=';', encoding='utf-8', index=False)
db5.to_csv('restoparts_product_type_withoutheader.csv', sep=';', encoding='utf-8', index=False, header=False)

In [108]:
db6.to_csv('restoparts_brand.csv', sep=';', encoding='utf-8', index=False)
db6.to_csv('restoparts_brand_withoutheader.csv', sep=';', encoding='utf-8', index=False, header=False)

In [109]:
db7.to_csv('restoparts_composition.csv', sep=';', encoding='utf-8', index=False)
db7.to_csv('restoparts_composition_withoutheader.csv', sep=';', encoding='utf-8', index=False, header=False)

In [110]:
db8.to_csv('restoparts_finish.csv', sep=';', encoding='utf-8', index=False)
db8.to_csv('restoparts_finish_withoutheader.csv', sep=';', encoding='utf-8', index=False, header=False)

In [111]:
db9.to_csv('restoparts_body_clean_exploded_composition.csv', sep=';', encoding='utf-8', index=False)
db9.to_csv('restoparts_body_clean_exploded_composition_withoutheader.csv', sep=';', encoding='utf-8', index=False, header=False)

In [112]:
db7.head(15)

ID_composition                                      Composition
0                0  steel production chromium steel 18/8 hot rolled
1                1                                    polypropylene
2                2                                rubber production
3                3                                  jute production
4                4                                  zinc production
5                5                  acrylonitrile-butadiene-styrene
6                6                                        polyamide
7                7                                reinforcing steel
8                8                          sheet rolling aluminium
9                9           steel production converted low-alloyed
10              10                          polyurethane production
11              11                   glass fibre reinforced plastic
12              12                       steel production converted

In [113]:
db8.head(10)

ID_finish                                          Finish
0          0                             electrostatic paint
1          1                             zinc coating pieces
2          2                                    black chrome
3          3                  seal production natural rubber
4          4                            powder coating steel
5          5                    chromium coat electroplating
6          6                                    plastic film
7          7  selective coating stainless steel black chrome
8          8                                      enamelling
9          9                 acrylonitrile-butadiene-styrene

In [114]:
composition_search = db9.loc[db9['ID_composition'] == ]  # put ID number
composition_search

SyntaxError: invalid syntax (2264434635.py, line 1)

In [ ]:
finish_search = db9.loc[db9['ID_finish'] == ]  # put ID number
finish_search